embedded-graphics/MIGRATING-0.5-0.6.md

```diff
@@ -0,0 +1,422 @@
+
+First, change the imports in `Cargo.toml`:
+
+```diff
+[dependencies]
+- embedded-graphics = { version = "0.5.1", features = [ "tga", "bmp" ] }
+
++ embedded-graphics = "0.6.0"
++ tinytga = { version = "0.3.0", features = [ "graphics" ] }
++ tinybmp = { version = "0.2.0", features = [ "graphics" ] }
+```
+
+Then any image code:
+
+```diff
+- use embedded_graphics::{prelude::*, image::ImageBmp};
+-
+- let image = ImageBmp::new(include_bytes!("../../../assets/patch_16bpp.bmp")).unwrap();
+-
+- display.draw(&image);
+
++ use embedded_graphics::{prelude::*, image::Image, geometry::Point}
++ use tinybmp::Bmp;
++
++ let bmp = Bmp::from_slice(include_bytes!("../../../assets/patch_16bpp.bmp")).unwrap();
++
++ let image = Image::new(&bmp, Point::zero());
++
++ image.draw(&mut display)?;
+```
+
+## For driver authors
+
+Adding embedded-graphics support to a display driver has changed somewhat. The `Drawing` trait has been replaced by the `DrawTarget` trait which provides a slightly different interface.
+
+An associated error type is also added to `DrawTarget` to allow for better error handling than a `panic!()`.
+
+Below is a reduced example taken from the [ssd1306](https://crates.io/crates/ssd1306) driver. It uses `BinaryColor` as pixels on the SSD1306 can only be on or off.
+
+```rust
+use embedded_graphics::{
+    drawable,
+    geometry::Size,
+    pixelcolor::{
+        raw::{RawData, RawU1},
+        BinaryColor, PixelColor,
+    },
+    DrawTarget,
+};
+
+impl<C> DrawTarget<C> for DisplayDriver
+where
+    C: PixelColor + Into<BinaryColor>,
+{
+    type Error = DI::Error;
+
+    fn draw_pixel(&mut self, pixel: drawable::Pixel<C>) -> Result<(), Self::Error> {
+        let drawable::Pixel(pos, color) = pixel;
+
+        // Guard against negative values. All positive i32 values from `pos` can be represented in
+        // the `u32`s that `set_pixel()` accepts...
+        if pos.x < 0 || pos.y < 0 {
+            return Ok(());
+        }
+
+        // ... which makes the `as` coercions here safe.
+        self.set_pixel(
+            pos.x as u32,
+            pos.y as u32,
+            // Convery color to BinaryColor, then into underlying raw storage u8
+            color.into().into_storage(),
+        );
+
+        Ok(())
+    }
+
+    fn size(&self) -> Size {
+        let (w, h) = self.get_dimensions();
+
+        Size::new(w as u32, h as u32)
+    }
+}
+```
+
+Some notes on the above:
+
+- The `Drawing` trait is renamed to `DrawTarget`.
+- This implementation takes a `C: PixelColor + Into<BinaryColor>` allowing items using other color types to be used on the monochrome display. If this is not possible or undesired, use the appropriate pixel type directly. For example:
```

Comment:
Ok, fair. I've reverted back to explicitly declaring `DrawTarget<PixelColor>`. This guide should serve as a description of patterns anyway, so that's good enough for now.